### PR TITLE
PSR-14 Event dispatcher

### PR DESCRIPTION
**Pull request by butschster**
| Q             | A
| ------------- | ---
| Bugfix?       | ❌
| Breaks BC?    | ✔️
| New feature?  | ✔️

This PR adds the ability to use the event system in the Spiral Framework.
The `Events` component was added, which allows registering listeners using the `configuration file` or the `Listener` attribute.

The PSR-14 EventDispatcher implementation must be implemented in a standalone package.

## Event

An event can be represented by a simple class:

```php
namespace Spiral\Router\Event;

use Spiral\Router\RouteInterface;

final class RouteFound
{
    public function __construct(
        public readonly RouteInterface $route
    ) {
    }
}
```

### Dispatching an event:

```php
$this->container->get(EventDispatcherInterface::class)->dispatch(new RouteNotFound($request));
```

## Listener

A listener can be represented by a simple class with a method that will be called to handle the event. The method name can be configured in the Listener attribute parameter or in the configuration file (` __invoke` by default):

```php
namespace App\Listener;

use Spiral\Events\Attribute\Listener;
use Spiral\Router\Event\RouteFound;

class RouteListener
{
    public function __invoke(RouteFound $event): void
    {
        // ...
    }
}
```

### Registering a listener via a config file:

```php

// file app/config/events.php

use App\Listener\RouteListener;
use Spiral\Events\Config\EventListener;
use Spiral\Router\Event\RouteFound;

return [
    'listeners' => [
        // without any options
        RouteFound::class => [
            RouteListener::class,
        ],

        // OR

        // with additional options
        RouteFound::class => [
            new EventListener(
                listener: RouteListener::class,
                method: 'onRouteFound',
                priority: 1 
            ),
        ],
    ]
];
```

### Registering a listener via an attribute:

The attribute can be used without additional parameters. Then the method name __invoke and the event from the type of the method parameter will be used:

```php
namespace App\Listener;

use Spiral\Events\Attribute\Listener;
use Spiral\Router\Event\RouteFound;

#[Listener]
class RouteListener
{
    public function __invoke(RouteFound $event): void
    {
        // ...
    }
}
```

All available options:

```php
namespace App\Listener;

use Spiral\Events\Attribute\Listener;
use Spiral\Router\Event\RouteFound;

#[Listener(event: RouteFound::class, method: 'onRouteFound', priority: 1)]
class RouteListener
{
    public function onRouteFound(RouteFound $event): void
    {
        // ...
    }
}
```

The attribute can be used directly on the method, then the method name can be omitted:

```php
namespace App\Listener;

use Spiral\Events\Attribute\Listener;
use Spiral\Router\Event\RouteFound;

class RouteListener
{
    #[Listener(priority: 1)]
    public function onRouteFound(RouteFound $event): void
    {
        // ...
    }
}
```

### Processors

By default, listeners can be configured with both ways via attributes and a configuration file. In the configuration file, you can change this behavior and use only one `Processor`:

```php
// file app/config/events.php

use Spiral\Events\Processor\AttributeProcessor;
use Spiral\Events\Processor\ConfigProcessor;

return [
    // by default
    'processors' => [
        AttributeProcessor::class,
        ConfigProcessor::class,
    ],

    // only via config file
    'processors' => [
        ConfigProcessor::class,
    ],
    
    // only via attributes
    'processors' => [
        AttributeProcessor::class,
    ],
];
```

## EventDispatcher implementation

We can use ready-made implementations and integrate them into the Spiral Framework.

An example of `league/event` package integration.

### ListenerRegistryInterface

Let's extend the `Spiral\Events\ListenerRegistryInterface` becouse the `league/event` package provides additional methods.

```php
namespace Spiral\League\Event;

use League\Event\ListenerSubscriber;
use Spiral\Events\ListenerRegistryInterface as SpiralListenerRegistryInterface;

interface ListenerRegistryInterface extends SpiralListenerRegistryInterface
{
    public function addOneTimeListener(string $event, callable $listener, int $priority = 0): void;

    public function addListenersFrom(ListenerSubscriber $subscriber): void;
}
```

And add the implementation:

```php
namespace Spiral\League\Event;

use League\Event\ListenerSubscriber;
use League\Event\PrioritizedListenerRegistry;
use Psr\EventDispatcher\ListenerProviderInterface;

final class ListenerRegistry implements ListenerRegistryInterface, ListenerProviderInterface
{
    public function __construct(
        private readonly PrioritizedListenerRegistry $registry
    ) {
    }

    public function addListener(string $event, callable $listener, int $priority = 0): void
    {
        $this->registry->subscribeTo($event, $listener, $priority);
    }

    public function addOneTimeListener(string $event, callable $listener, int $priority = 0): void
    {
        $this->registry->subscribeOnceTo($event, $listener, $priority);
    }

    public function getListenersForEvent(object $event): iterable
    {
        return $this->registry->getListenersForEvent($event);
    }

    public function addListenersFrom(ListenerSubscriber $subscriber): void
    {
        $this->registry->subscribeListenersFrom($subscriber);
    }
}
```

### Bootloader

Let's add a bootloader that will register the EventDispatcher and the ListenerRegistryInterface implementation in the container:

```php
namespace Spiral\League\Event\Bootloader;

use League\Event\EventDispatcher;
use Psr\EventDispatcher\EventDispatcherInterface;
use Spiral\Boot\Bootloader\Bootloader;
use Spiral\Events\Bootloader\EventsBootloader;
use Spiral\Events\ListenerRegistryInterface as SpiralListenerRegistryInterface;
use Spiral\League\Event\ListenerRegistry;
use Spiral\League\Event\ListenerRegistryInterface;

final class EventBootloader extends Bootloader
{
    protected const DEPENDENCIES = [
        EventsBootloader::class
    ];

    protected const SINGLETONS = [
        SpiralListenerRegistryInterface::class => ListenerRegistryInterface::class,
        ListenerRegistryInterface::class => ListenerRegistry::class,
        ListenerRegistry::class => ListenerRegistry::class,
        EventDispatcherInterface::class => [self::class, 'initDispatcher'],
        EventDispatcher::class => EventDispatcherInterface::class
    ];

    private function initDispatcher(ListenerRegistry $registry): EventDispatcher
    {
        return new EventDispatcher($registry);
    }
}
```

### CODE REVIEW

File: src/Boot/src/AbstractKernel.php [hunk 1]
```diff
@@ -92,7 +96,7 @@ public static function create(
         array $directories,
         bool $handleErrors = true,
         ExceptionHandlerInterface|string|null $exceptionHandler = null,
-        Container $container = new Container(),
+        Container $container = new Container()
```
Comment (roxblnfk) [inline src/Boot/src/AbstractKernel.php:99]:
```suggestion
        Container $container = new Container(),
```

File: src/Boot/src/AbstractKernel.php [hunk 2]
```diff
@@ -207,15 +215,25 @@ public function addDispatcher(DispatcherInterface $dispatcher): self
      */
     public function serve(): mixed
     {
+        $eventDispatcher = $this->container->has(EventDispatcherInterface::class) ?
+            $this->container->get(EventDispatcherInterface::class) :
+            null;
```
Comment (roxblnfk) [inline src/Boot/src/AbstractKernel.php:220]:
```suggestion
        $eventDispatcher = $this->container->has(EventDispatcherInterface::class)
            ? $this->container->get(EventDispatcherInterface::class)
            : null;
```

File: src/Boot/src/AbstractKernel.php [hunk 3]
```diff
@@ -207,15 +215,25 @@ public function addDispatcher(DispatcherInterface $dispatcher): self
...
+
         foreach ($this->dispatchers as $dispatcher) {
             if ($dispatcher->canServe()) {
                 return $this->container->runScope(
                     [DispatcherInterface::class => $dispatcher],
-                    static fn () => $dispatcher->serve()
+                    function () use ($dispatcher, $eventDispatcher): mixed {
```
Comment (roxblnfk) [inline src/Boot/src/AbstractKernel.php:226]:
```suggestion
                    static function () use ($dispatcher, $eventDispatcher): mixed {
```

File: src/Framework/Bootloader/Http/RouterBootloader.php [hunk 4]
```diff
@@ -69,7 +70,12 @@ private function router(
         UriHandler $uriHandler,
         ContainerInterface $container
     ): RouterInterface {
-        return new Router($this->config->getConfig(HttpConfig::CONFIG)['basePath'], $uriHandler, $container);
+        return new Router(
+            $this->config->getConfig(HttpConfig::CONFIG)['basePath'],
+            $uriHandler,
+            $container,
+            $container->has(EventDispatcherInterface::class) ? $container->get(EventDispatcherInterface::class) : null
+        );
```
Comment (roxblnfk) [inline src/Framework/Bootloader/Http/RouterBootloader.php:78]: You can use FactoryInterface instead

File: src/Router/src/Event/RouteFound.php [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Spiral\Router\Event;
+
+use Spiral\Router\RouteInterface;
+
+final class RouteFound
+{
+    public function __construct(
+        public readonly RouteInterface $route
+    ) {
```
Comment (roxblnfk) [inline src/Router/src/Event/RouteFound.php:13]: What about Request object here?

**Review by roxblnfk — CHANGES_REQUESTED** [review]

File: src/Boot/src/AbstractKernel.php (hunk 1 above)
Comment (msmakouz) [reply src/Boot/src/AbstractKernel.php]: Done.

File: src/Events/.travis.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+language: php
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Events/.travis.yml]: Removed.

File: src/Events/README.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# Spiral Framework: CSRF Middleware
+[![Latest Stable Version](https://poser.pugx.org/spiral/csrf/version)](https://packagist.org/packages/spiral/csrf)
+[![Build Status](https://travis-ci.org/spiral/csrf.svg?branch=master)](https://travis-ci.org/spiral/csrf)
+[![Codecov](https://codecov.io/gh/spiral/csrf/branch/master/graph/badge.svg)](https://codecov.io/gh/spiral/csrf/)
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Events/README.md]: Fixed.

**Comment by butschster** [comment]
- `Symfony\Component\Console\Application` has an ability to inject EventDispatcher via `Spiral\Console\Console`
- `Symfony\Component\Mailer\Mailer` has an ability to inject EventDispatcher via `Spiral\SendIt\Bootloader\MailerBootloader`

**Review by roxblnfk — CHANGES_REQUESTED** [review]

**Review by roxblnfk — CHANGES_REQUESTED** [review]
Please add descriptions for all events in `Spiral\Boot\Event`

**Review by roxblnfk — APPROVED** [review]

**Pull request merged**